### ISSUE TITLE
KeyError missing facet_counts while running indexer

### ISSUE THREAD

**Issue by nwoodward** [ilri/dspace-statistics-api#13]
I'm not sure how this scenario arises, but if there are no facet_counts in a page of the results, then the indexing process dies with the following error:

> Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/ilri/dspace-statistics-api/dspace_statistics_api/indexer.py", line 239, in <module>
    index_views("items", "id")
  File "/ilri/dspace-statistics-api/dspace_statistics_api/indexer.py", line 114, in index_views
    views = res.json()["facet_counts"]["facet_fields"]
KeyError: 'facet_counts'

It could be due to so many unmigrated IDs or something else, but I wonder if it makes sense to check for the key in the json response before referencing it on https://github.com/ilri/dspace-statistics-api/blob/v6_x/dspace_statistics_api/indexer.py#L114 and https://github.com/ilri/dspace-statistics-api/blob/v6_x/dspace_statistics_api/indexer.py#L196.

**Comment by alanorth** [comment]
Yeah that's a tricky one. I could check to see if `res.json()["facet_counts"]["facet_fields"]` exists, but by this point it is already assumed that there are results. A bit earlier on [line 70](https://github.com/ilri/dspace-statistics-api/blob/v6_x/dspace_statistics_api/indexer.py#L70) it checks to see that there are results:

```python
    try:
        # get total number of distinct facets (countDistinct)
        results_totalNumFacets = res.json()["stats"]["stats_fields"][facetField][
            "countDistinct"
        ]
    except TypeError:
        print(f"{indexType}: no views, exiting.")

        exit(0)
```

I had this issue a few times when I was still processing my stats. I can't remember why. Do you manage to index a few pages before it crashes, or is it at the beginning? Thanks for reporting and helping to debug this!

**Comment by nwoodward** [comment]
Sure, no problem! That's what's so strange. It's happening somewhere in the middle of the indexing. The last output from the process is:
> items: indexing views (page 196 of 1283)

As a test, I added a check for the existence of "facet_counts" in the response and continue in the loop if it's not there, and now the indexing process continues on after dozens of results pages where it's not there. More logs....

> items: indexing views (page 244 of 1283)
No facet_counts in json response
items: indexing views (page 245 of 1283)
No facet_counts in json response
items: indexing views (page 246 of 1283)
No facet_counts in json response
items: indexing views (page 247 of 1283)
No facet_counts in json response
items: indexing views (page 248 of 1283)
No facet_counts in json response
items: indexing views (page 249 of 1283)
No facet_counts in json response
items: indexing views (page 250 of 1283)
items: indexing views (page 251 of 1283)
items: indexing views (page 252 of 1283)
items: indexing views (page 253 of 1283)
items: indexing views (page 254 of 1283)
items: indexing views (page 255 of 1283)

**Comment by alanorth** [comment]
Huh! I didn't expect that. So I can definitely add a `continue` there to skip to the next iteration of the while loop when `facet_counts` is missing from the JSON response, but I'm really curious about what is in the Solr response in these cases. Can you print the URL for the failing page and try to check the Solr response manually? Something like this around line 113:

```python
                if not res.json()["facet_counts"]["facet_fields"]:
                    print(res.request.url)
                    continue
```

Thanks!

**Comment by nwoodward** [comment]
Yes, it's so strange! The indexing process goes well up until page 183 of 1283 and then facet_counts is missing for the next three pages (request.url's are below). Then it continues indexing fine for another ~10 pages, then facet_counts is missing again for ~20 pages, then it's back for a while, and the process repeats.  

>items: indexing views (page 183 of 1283)
No facet_counts in json response
http://localhost:8080/solr/statistics/select?q=type%3A2&fq=-isBot%3Atrue+AND+statistics_type%3Aview&fl=id&facet=true&facet.field=id&facet.mincount=1&facet.limit=100&facet.offset=18200&shards=http%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2014%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2015%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2017%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2018%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2019%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2020&rows=0&wt=json&json.nl=map
items: indexing views (page 184 of 1283)
No facet_counts in json response
http://localhost:8080/solr/statistics/select?q=type%3A2&fq=-isBot%3Atrue+AND+statistics_type%3Aview&fl=id&facet=true&facet.field=id&facet.mincount=1&facet.limit=100&facet.offset=18300&shards=http%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2014%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2015%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2017%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2018%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2019%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2020&rows=0&wt=json&json.nl=map
items: indexing views (page 185 of 1283)
No facet_counts in json response
http://localhost:8080/solr/statistics/select?q=type%3A2&fq=-isBot%3Atrue+AND+statistics_type%3Aview&fl=id&facet=true&facet.field=id&facet.mincount=1&facet.limit=100&facet.offset=18400&shards=http%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2014%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2015%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2017%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2018%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2019%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2020&rows=0&wt=json&json.nl=map
items: indexing views (page 186 of 1283)
items: indexing views (page 187 of 1283)
items: indexing views (page 188 of 1283)

When I check the Solr results manually, I don't see anything obvious. It starts with: 
`<response><lst name="responseHeader"><int name="status">0</int><int name="QTime">0</int><lst name="params"><str name="q">type:2</str></lst></lst><result name="response" numFound="16235" start="0"><doc>...`

And there's a lot of `<bool name="isBot">true</bool>` statements in the results. They all appear to be from bots. I can send the whole response if that would help.

**Comment by alanorth** [comment]
Yes that would be very helpful if you could capture a page of the results that exhibits this behavior:

```console
curl "http://localhost:8080/solr/statistics/select?q=type%3A2&fq=-isBot%3Atrue+AND+statistics_type%3Aview&fl=id&facet=true&facet.field=id&facet.mincount=1&facet.limit=100&facet.offset=18400&shards=http%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2014%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2015%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2017%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2018%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2019%2Chttp%3A%2F%2Flocalhost%3A8080%2Fsolr%2Fstatistics-2020&rows=0&wt=json&json.nl=map" > /tmp/response.json
```

I viciously purge all bot hits from our stats so we don't have any, but it shouldn't matter because the indexer is filtering them out with `-isBot:true`.

**Comment by nwoodward** [comment]
Hmm. I swear I was getting a 200 response and results that were all bots just a few days ago. When I run curl with any of the results pages sans "facet_counts" today I get back a 400.

> {"responseHeader":{"status":400,"QTime":25112,"params":{"q":"type:2","facet.limit":"100","shards":"http://localhost:8080/solr/statistics,http://localhost:8080/solr/statistics-2014,http://localhost:8080/solr/statistics-2015,http://localhost:8080/solr/statistics-2017,http://localhost:8080/solr/statistics-2018,http://localhost:8080/solr/statistics-2019,http://localhost:8080/solr/statistics-2020","facet.field":"id","json.nl":"map","fl":"id","fq":"-isBot:true AND statistics_type:view","facet.mincount":"1","rows":"0","facet":"true","wt":"json","facet.offset":"18400"}},"error":{"msg":"Expected mime type application/octet-stream but got application/xml. <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<response>\n<lst name=\"error\"><str name=\"msg\">application/x-www-form-urlencoded content length (2151935 bytes) exceeds upload limit of 2048 KB</str><int name=\"code\">400</int></lst>\n</response>\n","code":400}}

**Comment by alanorth** [comment]
OK that `exceeds upload limit` error is weird and distracting... but I also notice your query doesn't include the new UUID filter we added in #12:

```python
"q": f"type:2 AND {facetField}:/.{{36}}/"
```

**Comment by nwoodward** [comment]
OK, I've been testing the indexing process after updating the code to include the UUID filter, and so far I haven't hit another instance of missing facet_counts. So I think we can consider this solved. Thanks!

*One other benefit of the filter I've noticed is it reduced the number of results pages by more than half. I'm not sure if you saw in the results pasted above, but I used to have about 1,300 pages of results, and the indexing process took hours. Now the process finishes in about an hour.

**Comment by alanorth** [comment]
OK that is great. Thank you for the report and the tests! I will close this now.

<hr />

Regarding the indexing speed, I found that increasing the Solr `filterCache` helps. See [this commit](https://github.com/ilri/DSpace/commit/6af35d3cd6f21f07dff10d8972e4cad3d8786b30) in our DSpace repository and [my notes from 2019-04 about Solr caching](https://alanorth.github.io/cgspace-notes/2019-04/#2019-04-17). I should really do more tests now, but getting VisualVM wired up was a bit of a pain!

**Comment by nwoodward** [comment]
Do you have any advice for what I can do if I'm still receiving the `exceeds upload limit` even after including the UUID filter? I'm running into the issue on a different DSpace instance that's also quite large. I could add the continue statement, but I'm concerned that I'd be missing some statistics.

**Comment by alanorth** [comment]
@nwoodward I've never seen such an error here. We did hit a similar error in XMLUI, though. It seemed to be related to users who were in many groups, which caused the Solr queries to be excessively large. The solution was to [increase Tomcat's maxHttpHeaderSize](https://github.com/ilri/rmg-ansible-public/commit/358b5ea43f9e5820986f897c9d560937c702ac6e).

Is your "exceeds upload limit" error coming from Tomcat, Solr, or something else in front of them (like httpd)?

**Comment by nwoodward** [comment]
Thanks for the tip. I'll try adjusting our Tomcat config. I've run into this error in a few places, and like you said it has mostly been due to a lot of associated objects that get returned in the Solr results. E.g. I've added queries to get the number of items added to communities/collections in the last month and year similar to (1), and if one or more items has hundreds of bitstreams it would return `exceeds upload limit` until I added `rows:0` because I really only needed the `numFound` number.


1. https://wiki.lyrasis.org/display/DSPACE/Solr#Solr-Breakdownofsubmitteditemspermonth

**Comment by alanorth** [comment]
> like you said it has mostly been due to a lot of associated objects that get returned in the Solr results.

Our issue was that we really just couldn't even POST the Solr query because there were so many OR OR OR OR OR for all the groups. Your issue *might* be different then... at least you have a reliable way to trigger it for testing, it sounds like?

**Comment by nwoodward** [comment]
Sorry, I missed your question earlier. As to whether the error is from Apache, Tomcat, or Solr I'm not sure, but my best guess is Solr. When I searched around for the error text `application/x-www-form-urlencoded content length (2104547 bytes) exceeds upload limit of 2048 KB` I found recommendations to edit the `<requestParsers>` config in solrconfig.xml to increase the `formdataUploadLimitInKB` value.

I'll test this and the `maxHttpHeaderSize` param in the Tomcat config.
